File: Prufer Decoding/prufer_decoding.py
#This code is contributed by Riyazul555

# Time Complexity = O(N)  where N is number of elements in input array 
# Space Complexity = O(N) where N is number of elements in input array

def main():
    with open(sys.argv[1], 'r') as fin, open(sys.argv[2], 'w') as fout:
        n = int(fin.readline())
        cod = list(map(int, fin.readline().split()))
        deg = [0] * (n + 5)
        ptr = 0

        for i in range(n):
            deg[cod[i]] += 1

        while deg[ptr] != 0:
            ptr += 1

        leaf = ptr
        p = [-1] * (n + 5)

        for i in range(n):
            p[leaf] = cod[i]
            if deg[cod[i]] == 0 and cod[i] < ptr:
                leaf = cod[i]
            else:
                ptr += 1
                while deg[ptr] != 0:
                    ptr += 1
                leaf = ptr

        fout.write(str(n + 1) + "\n")
        fout.write(" ".join(map(str, p[:n+1])) + " ")

if __name__ == "__main__":
    import sys
    main()

File: Prufer Decoding/prufer_decoding.cpp
// This code is contributed by Riyazul555 and Iuliagroza

// Time Complexity = O(N)  where N is number of elements in input array 
// Space Complexity = O(N) where N is number of elements in input array

#include <fstream>
#include <vector>
#define nMax 100005

using namespace std;

int main(int argc, char* argv[])
{
    ifstream fin(argv[1]);
    ofstream fout(argv[2]);
    int n, cod[nMax], deg[nMax]={0};
    fin >> n;
    for(int i=0; i<n; i++)
    {
        fin >> cod[i];
        deg[cod[i]]++;
    }
    int ptr=0;
    while(deg[ptr]!=0)
        ptr++;
    int leaf=ptr;
    int p[nMax];
    for(int i=0; i<n; i++)
        p[i]=-1;
    for(int i=0; i<n; i++)
    {
        p[leaf]=cod[i];
        if(--deg[cod[i]]==0 && cod[i]<ptr)
            leaf=cod[i];
        else
        {
            ptr++;
            while(deg[ptr]!=0)
                ptr++;
            leaf=ptr;
        }
    }
    fout << n+1 << "\n";
    for(int i=0; i<n+1; i++)
        fout << p[i] << " ";
    fin.close();
    fout.close();
    return 0;
}